File: packages/dart/lib/src/session.dart
import 'dart:convert';

import 'config.dart';
import 'errors.dart';

/// Persisted account session (matches Rust/TS shape).
class AccountSession {
  AccountSession({
    required this.accountKey,
    this.accessToken,
    this.licenseJwt,
    this.licenseEtag,
    this.payingPartyIdHeader,
  });

  final String accountKey;
  String? accessToken;
  String? licenseJwt;
  String? licenseEtag;
  String? payingPartyIdHeader;

  Map<String, dynamic> toJson() => {
        'account_key': accountKey,
        'access_token': accessToken,
        'license_jwt': licenseJwt,
        'license_etag': licenseEtag,
        'paying_party_id_header': payingPartyIdHeader,
      };

  factory AccountSession.fromJson(Map<String, dynamic> json) {
    return AccountSession(
      accountKey: (json['account_key'] ?? json['accountKey']) as String,
      accessToken: (json['access_token'] ?? json['accessToken']) as String?,
      licenseJwt: (json['license_jwt'] ?? json['licenseJwt']) as String?,
      licenseEtag: (json['license_etag'] ?? json['licenseEtag']) as String?,
      payingPartyIdHeader:
          (json['paying_party_id_header'] ?? json['payingPartyIdHeader'])
              as String?,
    );
  }
}

/// Opaque key-value store (host supplies secure storage).
abstract class SessionStore {
  Future<String?> get(String key);
  Future<void> set(String key, String value);
  Future<void> delete(String key);
}

/// In-memory store for tests / CLI prototypes.
class MemorySessionStore implements SessionStore {
  final Map<String, String> _map = {};

  @override
  Future<String?> get(String key) async => _map[key];

  @override
  Future<void> set(String key, String value) async => _map[key] = value;

  @override
  Future<void> delete(String key) async => _map.remove(key);
}

String _sanitize(String accountKey) =>
    accountKey.replaceAll(RegExp(r'[^a-zA-Z0-9_-]'), '_');

class SessionManager {
  SessionManager({
    required TwoKeySdkConfig config,
    SessionStore? store,
  })  : _config = config.validated(),
        _store = store ?? MemorySessionStore();

  final TwoKeySdkConfig _config;
  final SessionStore _store;

  String _key(String accountKey) =>
      '${_config.storagePrefix}:session:${_sanitize(accountKey)}';

  Future<AccountSession?> load(String accountKey) async {
    final raw = await _store.get(_key(accountKey));
    if (raw == null) return null;
    try {
      return AccountSession.fromJson(
        jsonDecode(raw) as Map<String, dynamic>,
      );
    } catch (e) {
      throw TwoKeyException(
        TwoKeyErrorCode.unknown,
        'Corrupt session data',
        detail: '$e',
      );
    }
  }

  Future<void> save(AccountSession session) async {
    await _store.set(_key(session.accountKey), jsonEncode(session.toJson()));
  }

  Future<void> clear(String accountKey) async {
    await _store.delete(_key(accountKey));
  }
}

File: packages/ts/src/session.ts
import type { SdkConfig } from "./config.js";
import { validateConfig } from "./config.js";
import { TwoKeyError } from "./errors.js";

/** Minimal persisted account session (browser). */
export type AccountSession = {
  accountKey: string;
  accessToken?: string;
  licenseJwt?: string;
  licenseEtag?: string;
  payingPartyIdHeader?: string;
};

export type SessionStore = {
  get(key: string): string | null | Promise<string | null>;
  set(key: string, value: string): void | Promise<void>;
  delete(key: string): void | Promise<void>;
};

/** localStorage-backed store (browser). */
export function localStorageSessionStore(): SessionStore {
  return {
    get(key) {
      return globalThis.localStorage?.getItem(key) ?? null;
    },
    set(key, value) {
      globalThis.localStorage?.setItem(key, value);
    },
    delete(key) {
      globalThis.localStorage?.removeItem(key);
    },
  };
}

/** In-memory store (tests / SSR stubs). */
export function memorySessionStore(): SessionStore {
  const map = new Map<string, string>();
  return {
    get: (k) => map.get(k) ?? null,
    set: (k, v) => {
      map.set(k, v);
    },
    delete: (k) => {
      map.delete(k);
    },
  };
}

function sanitize(accountKey: string): string {
  return accountKey.replace(/[^a-zA-Z0-9_-]/g, "_");
}

export class BrowserSessionManager {
  private readonly config: SdkConfig;
  private readonly store: SessionStore;

  constructor(config: SdkConfig, store: SessionStore = memorySessionStore()) {
    this.config = validateConfig(config);
    this.store = store;
  }

  private key(accountKey: string): string {
    return `${this.config.storagePrefix}:session:${sanitize(accountKey)}`;
  }

  async load(accountKey: string): Promise<AccountSession | null> {
    const raw = await this.store.get(this.key(accountKey));
    if (!raw) return null;
    try {
      return JSON.parse(raw) as AccountSession;
    } catch (e) {
      throw new TwoKeyError("unknown", "Corrupt session data", String(e));
    }
  }

  async save(session: AccountSession): Promise<void> {
    await this.store.set(this.key(session.accountKey), JSON.stringify(session));
  }

  async clear(accountKey: string): Promise<void> {
    await this.store.delete(this.key(accountKey));
  }
}
